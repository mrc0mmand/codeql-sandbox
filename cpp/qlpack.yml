name: codeql/cpp-tests
version: 0.0.1
dependencies:
  codeql/cpp-all: "*"
  codeql/cpp-queries: "*"
extractor: cpp
tests: .
