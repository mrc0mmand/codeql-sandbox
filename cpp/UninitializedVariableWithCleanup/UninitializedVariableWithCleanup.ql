/**
 * vi: sw=2 ts=2 et syntax=ql:
 *
 * Based on cpp/uninitialized-local.
 *
 * @name Potentially uninitialized local variable using the cleanup attribute
 * @description Running the cleanup handler on a possibly uninitialized variable
 *              is generally a bad idea.
 * @id cpp/uninitialized-local-with-cleanup
 * @kind problem
 * @problem.severity error
 * @precision high
 * @tags security
 */

import cpp
import semmle.code.cpp.controlflow.StackVariableReachability

/** Auxiliary predicate: List cleanup functions we want to explicitly ignore
  * since they don't do anything illegal even when the variable is uninitialized
  */
predicate cleanupFunctionDenyList(string fun) {
  fun = "erase_char"
}

/**
 * A declaration of a local variable using __attribute__((__cleanup__(x)))
 * that leaves the variable uninitialized.
 */
DeclStmt declWithNoInit(LocalVariable v) {
  result.getADeclaration() = v and
  not v.hasInitializer() and
  /* The variable has __attribute__((__cleanup__(...))) set */
  v.getAnAttribute().hasName("cleanup") and
  /* Check if the cleanup function is not on a deny list */
  not cleanupFunctionDenyList(v.getAnAttribute().getAnArgument().getValueText())
}

class UninitialisedLocalReachability extends StackVariableReachability {
  UninitialisedLocalReachability() { this = "UninitialisedLocal" }

  override predicate isSource(ControlFlowNode node, StackVariable v) { node = declWithNoInit(v) }

  /* Note: _don't_ use the `useOfVarActual()` predicate here (and a couple of lines
   * below), as it assumes that the callee always modifies the variable if
   * it's passed to the function.
   *
   * i.e.:
   * _cleanup_free char *x;
   * fun(&x);
   * puts(x);
   *
   * `useOfVarActual()` won't treat this an an uninitialized read even if the callee
   * doesn't modify the argument, however, `useOfVar()` will
   */
  override predicate isSink(ControlFlowNode node, StackVariable v) { useOfVar(v, node) }

  override predicate isBarrier(ControlFlowNode node, StackVariable v) {
    // only report the _first_ possibly uninitialized use
    useOfVar(v, node) or
    (
      /* If there's a return statement somewhere between the variable declaration
       * and a possible definition, don't accept is as a valid initialization.
       *
       * E.g.:
       * _cleanup_free_ char *x;
       * ...
       * if (...)
       *    return;
       * ...
       * x = malloc(...);
       *
       * is not a valid initialization, since we might return from the function
       * _before_ the actual iniitialization (emphasis on _might_, since we
       * don't know if the return statement might ever evaluate to true).
       */
      definitionBarrier(v, node) and
      not exists(ReturnStmt rs | v.getFunction() = rs.getEnclosingFunction() | rs.getLocation().isBefore(node.getLocation()))
    )
  }
}

from UninitialisedLocalReachability r, LocalVariable v, VariableAccess va
where r.reaches(_, v, va)
select va, "The variable $@ may not be initialized here, but has a cleanup handler.", v, v.getName()
