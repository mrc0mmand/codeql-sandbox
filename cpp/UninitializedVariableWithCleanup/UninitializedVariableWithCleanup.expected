| test.c:54:14:54:43 | b_macro_attribute_and_fun_call | The variable $@ may not be initialized here, but has a cleanup handler. | test.c:37:26:37:55 | b_macro_attribute_and_fun_call | b_macro_attribute_and_fun_call |
| test.c:58:10:58:25 | b_full_attribute | The variable $@ may not be initialized here, but has a cleanup handler. | test.c:33:45:33:60 | b_full_attribute | b_full_attribute |
| test.c:59:10:59:28 | b_macro_attribute_1 | The variable $@ may not be initialized here, but has a cleanup handler. | test.c:35:25:35:43 | b_macro_attribute_1 | b_macro_attribute_1 |
| test.c:61:10:61:34 | b_stack_struct_field_free | The variable $@ may not be initialized here, but has a cleanup handler. | test.c:39:46:39:70 | b_stack_struct_field_free | b_stack_struct_field_free |
| test.c:88:10:88:21 | b_unassigned | The variable $@ may not be initialized here, but has a cleanup handler. | test.c:78:26:78:37 | b_unassigned | b_unassigned |
